test/chat_models/chat_ollama_ai_test.exs

```diff
@@ -0,0 +1,224 @@
+      assert {"must be less than or equal to %{number}", _} = changeset.errors[:temperature]
+      assert {"must be less than or equal to %{number}", _} = changeset.errors[:mirostat_eta]
+    end
+
+    test "supports overriding the API endpoint" do
+      override_url = "http://localhost:99999/api/chat"
+
+      model =
+        ChatOllamaAI.new!(%{
+          endpoint: override_url
+        })
+
+      assert model.endpoint == override_url
+    end
+  end
+
+  describe "for_api/3" do
+    setup do
+      {:ok, ollama_ai} =
+        ChatOllamaAI.new(%{
+          "model" => "llama2:latest",
+          "temperature" => 0.4,
+          "stream" => false,
+          "seed" => 0,
+          "num_ctx" => 2048,
+          "num_predict" => 128,
+          "repeat_last_n" => 64,
+          "repeat_penalty" => 1.1,
+          "mirostat" => 0,
+          "mirostat_eta" => 0.1,
+          "mirostat_tau" => 5.0,
+          "num_gqa" => 8,
+          "num_gpu" => 1,
+          "num_thread" => 0,
+          "receive_timeout" => 300_000,
+          "stop" => "",
+          "tfs_z" => 0.0,
+          "top_k" => 0,
+          "top_p" => 0.0
+        })
+      %{ollama_ai: ollama_ai}
+    end
+
+    test "generates a map for an API call with no messages", %{ollama_ai: ollama_ai} do
+      data = ChatOllamaAI.for_api(ollama_ai, [], [])
+      assert data.model == "llama2:latest"
+      assert data.temperature == 0.4
+      assert data.stream == false
+      assert data.messages == []
+      assert data.seed == 0
+      assert data.num_ctx == 2048
+      assert data.num_predict == 128
+      assert data.repeat_last_n == 64
+      assert data.repeat_penalty == 1.1
+      assert data.mirostat == 0
+      assert data.mirostat_eta == 0.1
+      assert data.mirostat_tau == 5.0
+      assert data.num_gqa == 8
+      assert data.num_gpu == 1
+      assert data.num_thread == 0
+      assert data.receive_timeout == 300_000
+      # TODO: figure out why this is field is is being cast to nil instead of
+      assert data.stop == nil
+      assert data.tfs_z == 0.0
+      assert data.top_k == 0
+      assert data.top_p == 0.0
+    end
+
+    test "generates a map for an API call with a single message", %{ollama_ai: ollama_ai} do
+      user_message = "What color is the sky?"
+
+      data = ChatOllamaAI.for_api(ollama_ai, [Message.new_user!(user_message)], [])
+      assert data.model == "llama2:latest"
+      assert data.temperature == 0.4
+      assert get_in(data, [:messages, Access.at(0), "role"]) == :user
+      assert get_in(data, [:messages, Access.at(0), "content"]) == "What color is the sky?"
+    end
+
+    test "generates a map for an API call with user and system messages", %{ollama_ai: ollama_ai} do
+      user_message = "What color is the sky?"
+      system_message = "You are a weather man"
+
+      data =
+        ChatOllamaAI.for_api(
+          ollama_ai,
+          [Message.new_user!(user_message), Message.new_system!(system_message)],
```

Comment:
Is this intentionally in reverse order? Typically the system message is expected to be first and user messages follow.

Instead of using procedural `get_in` calls with Access.at/1 calls, pattern matches are more clear.  For example:

```elixir
assert [system_msg, user_msg] = data.messages
assert system_msg.role == :system
```
This fetches the messages and asserts a list of 2 elements all at once.

Or even like this: 
```elixir
assert [%Message{role: :system} = system_msg, %Message{role: :user} = user_msg] = data.messages
assert system_msg.content == "You are a weather man"
assert user_msg.content == "What color is the sky?"
```

Where we can assert 2 messages received in a particular order and the roles are asserted at that time. Then the longer content can be asserted separately.

Reply:
This makes sense. I am still pretty new to elixir so this is helpful!